spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+# before being copied to stdout when the group has completed
+PARALLEL_LOGGING_DIR_NAME = "parrTmp"
+
+# The maximum permissible width of divider text describing a parallel command group
+MAX_WIDTH_DIVIDER = 60
+
+# Whether or not to display realtime status information
+DISPLAY_STATUS_TABLE = sys.stdout.isatty() and supports_ansi()
+
+
+class _ParallelCommandState:
+    def __init__(
+        self,
+        name: str,
+        display_color: str,
+        # transitions: the names of states that can legally follow this state
+        transitions: Optional[List[str]] = None,
+    ):
+        self.name = name
+        self.display_color = display_color
+        if transitions is not None:
+            self.transitions = transitions
+        else:
+            self.transitions = []
+
+
+_PARALLEL_COMMAND_STATES = (
+    # The command has not yet been run
+    _ParallelCommandState("pending", "yellow", ["starting", "cancelled", "not rerun"]),
+    # The command was not run because its settings, inputs and outputs had not changed
+    # since a previous successful run
+    _ParallelCommandState("not rerun", "blue"),
+    # The main process has spawned a worker process for the command but not yet received
+    # an acknowledgement
+    _ParallelCommandState("starting", "green", ["running", "failed", "hung"]),
+    # The command is running
+    _ParallelCommandState(
+        "running", "green", ["running", "succeeded", "failed", "terminated", "hung"]
+    ),
+    # The command succeeded (rc=0)
+    _ParallelCommandState("succeeded", "green"),
+    # The command failed (rc>0)
+    _ParallelCommandState("failed", "red"),
+    # The command was terminated (rc<0), usually but not necessarily by the main process because
+    # another command within the same parallel group failed
+    _ParallelCommandState("terminated", "red"),
+    # The main process would expect the worker process to be running, but the worker process has
+    # not communicated with the main process for an extended period of time
+    _ParallelCommandState("hung", "red"),
+    # The main process did not attempt to execute the command because another command in the same
+    # parallel group had already failed
+    _ParallelCommandState("cancelled", "red"),
+)
+
+
+class _ParallelCommandInfo:
+
+    state_dict = {state.name: state for state in _PARALLEL_COMMAND_STATES}
+
+    def __init__(self, cmd_name: str, cmd: Dict, cmd_index: int):
+        self.cmd_name = cmd_name
+        self.cmd = cmd
+        self.cmd_index = cmd_index
+        self.state = _ParallelCommandInfo.state_dict["pending"]
+        self.pid: Optional[int] = None
+        self.last_keepalive_time: Optional[int] = None
+        self.running_os_cmd_index: Optional[int] = None
+        self.rc: Optional[int] = None
+        self.console_output: Optional[str] = None
+
+    def change_state(self, new_state: str):
+        assert new_state in self.state.transitions, (
```

Comment:
Not sure about the convention, but I think we want all errors to be properly defined error codes instead of `asserts`.